cache-manager/cache-manager.php

```diff
@@ -0,0 +1,433 @@
+
+			curl_multi_remove_handle( $curl_multi, $completed['handle'] );
+		}
+
+		curl_multi_close( $curl_multi );
+	}
+
+	protected function build_purge_request( $url, $method ) {
+		global $varnish_servers;
+
+		$requests = array();
+
+		if ( empty( $varnish_servers ) )
+			return $requests;
+
+		$parsed = parse_url( $url );
+		if ( empty( $parsed['host'] ) )
+			return $requests;
+
+		foreach ( $varnish_servers as $server  ) {
+			$server = explode( ':', $server[0] );
+
+			$uri = '/';
+			if ( isset( $parsed['path'] ) )
+				$uri = $parsed['path'];
+			if ( isset( $parsed['query'] ) )
+				$uri .= $parsed['query'];
+
+			$requests[] = array(
+				'ip'     => $server[0],
+				'port'   => $server[1],
+				'host'   => $parsed['host'],
+				'uri'    => $uri,
+				'method' => $method
+			);
+		}
+
+		return $requests;
+	}
+
+	public function execute_purges() {
+		$this->ban_urls = array_unique( $this->ban_urls );
+		$this->purge_urls = array_unique( $this->purge_urls );
+
+		if ( empty( $this->ban_urls ) && empty( $this->purge_urls ) )
+			return;
+
+		$requests = array();
+		foreach( (array) $this->ban_urls as $url )
+			$requests = array_merge( $requests, $this->build_purge_request( $url, 'BAN' ) );
+
+		foreach( (array) $this->purge_urls as $url )
+			$requests = array_merge( $requests, $this->build_purge_request( $url, 'PURGE' ) );
+
+		$this->ban_urls = $this->purge_urls = array();
+
+		if ( empty( $requests ) )
+			return;
+
+		return $this->curl_multi( $requests );
+	}
+
+	public function purge_site_cache( $when = null ) {
+		if ( $this->site_cache_purged )
+			return;
+
+		$this->ban_urls[] = untrailingslashit( home_url() ) . '/.*';
+		$this->site_cache_purged = true;
+
+		return;
+	}
+
+	public function queue_post_purge( $post_id ) {
+		if ( $this->site_cache_purged )
+			return false;
+
+		if ( defined( 'WP_IMPORTING' ) ) {
+			$this->purge_site_cache();
+			return false;
+		}
+
+		$post = get_post( $post_id );
+		if ( empty( $post ) ||
+		     'revision' === $post->post_type ||
+		     ! in_array( get_post_status( $post_id ), array( 'publish', 'trash' ), true ) )
+		{
+			return false;
+		}
+
+		$this->purge_urls[] = get_permalink( $post_id );
+		$this->purge_urls[] = trailingslashit( home_url() );
+
+		$taxonomies = get_object_taxonomies( $post, 'object' );
+
+		foreach ( $taxonomies as $taxonomy ) {
+			if ( true !== $taxonomy->public ) {
+				continue;
+			}
+			$taxonomy_name = $taxonomy->name;
+			$terms = get_the_terms( $post_id, $taxonomy_name );
+			if ( false === $terms ) {
+				continue;
+			}
+			foreach ( $terms as $term ) {
+				$this->queue_purge_urls_for_term( $term );
+			}
+		}
+
+		$feeds = array(
+			get_bloginfo('rdf_url'),
+			get_bloginfo('rss_url') ,
+			get_bloginfo('rss2_url'),
+			get_bloginfo('atom_url'),
+			get_bloginfo('comments_atom_url'),
+			get_bloginfo('comments_rss2_url'),
+			get_post_comments_feed_link( $post_id )
+		);
+
+		foreach ( $feeds as $feed ) {
```

Comment:
Addressed in https://github.com/Automattic/vip-go-mu-plugins/pull/247/commits/ce393a9dfc8ec92da61dfad361fd6c2fbbe75a8b